# 根据过去的行为能否预测当下（逻辑回归分析） 

## 1 背景介绍 

### 1.1 从非智能手机更换到智能手机

我们怀疑很多《垂钓乐园》游戏的用户从非智能手机更换到智能手机后，过去使用的游戏账号可能不能继续在新的手机上使用了。因此需要**根据之前用户的访问情况来预估一下到底有多少用户因为这个原因流失了，**那么应该怎么做？

### 1.2 《垂钓乐园》游戏用户量减少的问题

除了《黑猫拼图》游戏，公司还有一款名为《垂钓乐园》的社交游戏。这个游戏开始投入市场的时候**智能手机还没出现**，因此当时这款游戏是针对非智能手机定制开发的。然而，随着智能手机的快速普及，公司也随之在这款游戏中加如了针对智能手机的功能，以及用户账号的**迁转功能**，以便用户在智能手机上**继续使用过去非智能手机中的游戏账号**。这些功能上线后，智能手机的游戏用户稳步上升，同时游戏的总用户量也在不断增加。  

然而，**最近《垂钓乐园》的用户量却在不断减少**。我们从各种角度分析了这个问题，然后发现**游戏的非智能手机用户大量减少了**。

### 1.3 非智能手机用户量的减少主要由移动终端市场的变化所致

首先，随着移动终端市场的发展，非智能手机的用户量在逐渐减少，而智能手机的用户量在不断增加。从这样的趋势来看，使用非智能手机的游戏用户不断减少是**大势所趋**。其它游戏的非智能手机用户的情况也是如此。但是，在其它游戏中，虽然**非智能手机用户量在减少，但是是能手机用户量在增加，因而总用户量大致没有变化**。也就是说，和其它游戏相比，**《垂钓乐园》游戏的非智能手机用户量也减少了，但是智能手机用户量的增加幅度却比较小**。

## 2 现状和预期 

希望用智能手机用户量的增加来弥补非智能手机用户量的减少，所以：
- **现状：**智能手机用户量的增加比非智能手机用户量减少量要**少**
- **预期：**用智能手机用户量的增加来弥补非智能手机用户量的减少

## 3 发现问题

### 3.1 非智能手机用户流失的结构 

为什么游戏的智能手机用户量的增加会比非智能手机用户量的减少要小呢？这需要仔细思考非智能手机用户流失的结构。我们认为这些流失的用户大致可分为以下两部分：
- **一直使用非智能手机的用户流失：**这些用户是自然流失，而不是因为更换了智能手机（就是不想玩了）
- **从非智能手机更换为智能手机时的用户流失：**
    - 游戏账号迁移后继续在智能手机上使用
    - 在智能手机上新申请一个游戏账号
    - 干脆不玩了
    
将游戏账号从非智能手机迁移到智能手机的用户，将其算作是智能手机用户。但是，有些用户在更换智能手机后就重新申请了一个游戏账号。由于这些新申请账号的用户在数据上显示为**新用户**，因此很难判断这样的用户是不是更换了手机（非智能手机也可能注册新用户）。但是从定期的问卷调查来看，**只有极少数用户会在更换智能手机后重新申请一个游戏账号**。最后一种情况是用户在更换了智能手机后就不再玩《垂钓乐园》游戏了，这类用户和那些仍在使用非智能手机且自然流失的用户很难区分开。

以上内容整理后如下所示：

|游戏流失用户是否持有智能手机？|非智能手机用户流失原因|
|---|---|
|没有|自然流失|
|有|自然流失|
|有|在智能手机上使用了新的游戏账号|
|有|在智能手机上将旧的游戏账号迁转设定成功|

### 3.2 游戏账号迁移设定失败导致用户流失的假设

基于上表内容，经过和《垂钓乐园》游戏的策划和开发负责人开会商讨，怀疑在那些因更换智能手机而流失的用户当中，有的**可能是因为旧的游戏账号在智能手机上迁移设定失败**。游戏账号迁移设定模块当时开发的比较仓促，对用户来说有些使用不便。但是当时这个模块在发布后并没有出现什么问题，这可能时因为当时使用智能手机的用户大都属于时尚认识，就算游戏账号的迁移设定有点复杂，他们也能顺利完成。

将上面内容添加到表里，如下所示：

|游戏流失用户是否持有智能手机？|非智能手机用户流失原因|
|---|---|
|没有|自然流失|
|有|自然流失|
|有|在智能手机上使用了新的游戏账号|
|有|在智能手机上将旧的游戏账号迁转设定成功|
|有|在智能手机上将旧的游戏账号迁转审定失败|

如果所面临的本质问题时上表中最后的“**游戏账号迁移设定失败**”的画，就必须尽快修复这个问题。然而，由于这个模块当时开发的比较仓促，因此现在修复起来预计要消耗相当多的工时（**代价大**）。而现在开发部门也有很多开发任务，很难抽出时间。

因此，现在做问卷调查可能不太合适，我们需要的时一个大致的结论，并尽快预估游戏账号迁移设定失败造成的影响，进而采取下述某个解决对策：
- **如果游戏账号迁移设定失败造成了很大的影响，那就需要开发部门停止正在进行的工作，全力修复这个问题**
- **如果这个影响较小，就只需要改进以下针对智能手机用户的商业宣传**

## 4 数据收集 

### 4.1 用户自然流失和账号迁移设定失败导致的用户流失有何不同

首先，整理一下目前为止的事实和假设：
- **非智能手机用户量的减少比智能手机用户量的增加要大（事实）**
- **当前用户从非智能手机更换到智能手机后，重新注册游戏账号的用户很少（事实，从问卷调查中得出的结论）**
- **从非智能手机更换到智能手机时，游戏账号的迁移设定导致了用户流失（假设）**

为了验证上述假设，需要能够**在数据上区分用户的自然流失和游戏账号迁移设定失败而导致的用户流失**。那么什么样的数据才能满足分析需要呢？先看一下用户在流失之前的访问数据。如下表所示，自然流失的用户的访问天数时在逐渐减少的，而因游戏账号迁移设定失败而流失的用户是在某天后就突然不访问了。

|访问日期|1|2|3|4|5|6|7|8|
|---|---|---|---|---|---|---|---|---|
|自然流失的用户|X|O|X|X|O|X|X|X|
因游戏账号迁移设定失败而流失的用户|O|O|O|O|O|X|X|X|


这个假设虽然不是来自与问卷调查或者访谈，但也符合我们自身使用的感觉和游戏模块开发者的亲身体会，因此具有较高的可信度。

我们使用了2013年1越和2越的DAU数据。具体来说，就是**用1月份用户的访问数据来预测用户在2月份是自然流失，还是因游戏账号迁移设定失败而流失的**。

### 4.2 不存在包含正解的数据

为了判断用户是自然流失还是因游戏账号迁移设定失败而流失的，将用户1月份访问情况设置为自变量，这从自身的角度和游戏开饭人员的角度来看都是没有问题的。然后将“自然流失”和“因账号迁移设定失败而流失”作为因变量，并基于之前的表格来思考。

用户流失之前的访问情况大致可概括为以下几个特征：
- **访问次数较少：**
    - 未使用智能手机的用户自然流失
    - 使用智能手机的用户自然流失
- **访问次数较多：**
    - 在智能手机上使用了新的游戏账号
    - 在智能手机上将旧的游戏账号迁转设定成功
    - 在智能手机上将旧的游戏账号迁转审定失败
    
如上所示，如果访问情况存在显著差异，那么就能构筑判别模型。但是，能否梳理获得上面这些访问数据呢？下表给出相关整理情况：

|用户流失时是否持有智能手机？|非智能手机用户流失的原因|数据是否存在？|
|---|---|---|
|没有|自然流失（没有分析价值）|X|
|有|自然流失（没有分析价值）|X|
|有|在智能手机上使用了新的游戏账号|X（极少）|
|有|在智能手机上将旧的游戏账号迁转设定成功|O|
|有|在智能手机上将旧的游戏账号迁转设定失败|X|

由上表可知，所有的非智能手机的流失用户当中，之后的访问情况能够识别出来的只有第四行的“在智能手机上将旧的游戏账号迁转设定成功”这一种。第3行的“在智能手机上使用了新的游戏账号”的用户量很少，而最想知道的“在智能手机上将旧的的游戏账号迁转审定失败”这种情况的相关数据实际上无法获得。也就是说，**用户“自然流失”和“因游戏账号迁转设定失败而流失”这两种情况在数据上无法区分，无法获得所包含正解的数据**。

### 4.3 在无法获得包含正解的数据的情况下进行建模

在商业领域中，建模所需的数据不一定要局限于包含正解的完美数据。在本例中：
- 假设“因游戏账号迁转设定失败而流失”情况很严重：参考之前的表格，可以预知，和自然流失用户相比，因游戏账号迁转设定失败而流失的用户在当月的访问次数要高的多。因此，如果“因游戏账号迁转失败而流失”的用户在所流失用户中所占的比例较高的话，那么**流失用户和游戏账号迁转用户在1月份的访问情况就不会存在什么差异**，而如果访问情况没有什么差异的话，就无法基于访问数量构筑判别模型。
- 反之，如果“因游戏账号迁转设定失败而流失”的用户较少的话，那么访问次数的差异就会比较显著，也就能够顺利的构筑判别模型。在本例中，如果能构筑判别模型，就意味着令人畏惧的游戏账号迁转设定的复杂性并未产生影响。

### 4.4 分析过程整理 

整理一下整个分析过程：
- **事实：**
    - 非智能手机用户的减少比智能手机用户量的增加要大
    - 当用户从非智能手机更换到智能手机后，重新注册游戏账号的用户很少（从问卷调查中得出的结论）
- **假设：**
    - 从非智能手机更换到智能手机时，游戏账号的迁转设定失败导致了用户流失
- **根据分析方法和结果得出的解决对策：**
    - 使用用户访问情况数据建立“账号迁移用户”和“流失用户”的判别模型：
        - **能够建立有效的模型：** 因游戏账号迁转设定失败而导致的用户流失的影响较小
        - **不能建立模型：**很可能时因游戏账号迁转设定失败而导致的用户流失的影响较大，可以考虑简化游戏账号的迁转设定

## 5 数据收集和加工

对于本例中，需要建立“账号迁转用户”和“流失用户”的判别模型。对于这种二分类问题，有很多模型能够满足需求，但是在本例中，由于没有包含正解的数据，因此没有必要使用注入SVM（支持向量机）或者神经网络这类精度虽高但是计算量很大的模型，不然使用逻辑回归分析尽快得出结果。

### 5.1 数据处理和可视化准备

In [549]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
%matplotlib inline
plt.rcParams['font.sans-serif'] = ['SimHei']
plt.rcParams['axes.unicode_minus'] = False

### 5.2 数据收集

本例中，使用的是2013年1月份和2月份的数据

In [550]:
# 加载数据
dau = pd.read_csv('./data/ch7/section7-dau.csv')

In [551]:
dau.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48988 entries, 0 to 48987
Data columns (total 5 columns):
region_month    48988 non-null object
region_day      48988 non-null object
app_name        48988 non-null object
user_id         48988 non-null int64
device          48988 non-null object
dtypes: int64(1), object(4)
memory usage: 1.9+ MB


In [552]:
dau.head()

region_month  region_day app_name   user_id device
0      2013-01  2013-01-01  game-02  10061580     FP
1      2013-01  2013-01-01  game-02  10154440     FP
2      2013-01  2013-01-01  game-02  10164762     SP
3      2013-01  2013-01-01  game-02  10165615     FP
4      2013-01  2013-01-01  game-02  10321356     FP

In [553]:
dau.region_month.value_counts()

2013-01    25847
2013-02    23141
Name: region_month, dtype: int64

通过上述可知，一共有接近5万条数据，其中数据完整没有缺失值。1月份的数据有25847条，2月份的数据有23141条。

各属性解释：
- region_month：登录月份
- region_day：登录日期
- app_name：应用名称
- user_id：用户id
- device：设备类型（FP为非智能设备，SP为智能设备）

### 5.3 数据加工

逻辑回归分析所需的数据，如下所示：
- **关于用户是否迁移到了其它账号的数据**
- **关于用户是否每天访问游戏的数据（如下表所示）**

|用户ID|1日|2日|······|31日|
|---|---|---|---|---|
|100|1|0|······|1|
|101|1|1|······|0|
|···|···|···|······|···|

基于上表的数据，将其加工成适用于逻辑回归分析的数据形式。

#### 5.3.1 关于用户是否迁移到了其它账号的数据整理 

首先，生成关于1月份非智能手机用户在2月份是否依然访问游戏的数据。

用户登录情况：
- 同一个月中，不同日期登录
- 不同月中，不同日期登录
- 不同月中，不同设备登录

**按登录月份、用户ID、设备类型来统计用户**

In [554]:
mau = dau.drop_duplicates(subset=['region_month', 'user_id', 'device'])[['region_month', 'user_id', 'device']]

In [555]:
mau.head(3)

region_month   user_id device
0      2013-01  10061580     FP
1      2013-01  10154440     FP
2      2013-01  10164762     SP

**从mau中提取1月和2月非智能手机用户数据**

In [556]:
fp_mau = mau[mau.device=='FP']
fp_mau_1 = fp_mau.loc[fp_mau.region_month=='2013-01'].copy()
fp_mau_2 = fp_mau.loc[fp_mau.region_month=='2013-02'].copy()

**从mau中提取1月和2月智能手机用户数据**

In [557]:
sp_mau = mau[mau.device=='SP']
sp_mau_1 = sp_mau.loc[sp_mau.region_month=='2013-01'].copy()
sp_mau_2 = sp_mau.loc[sp_mau.region_month=='2013-02'].copy()

**Tips：**以上切片后使用copy()函数，是为了避免后面增加标志位时产生的**链式赋值**而产生的警告。

**1月份非智能手机用户在2月份的访问情况**

In [558]:
# 给mau添加一个is_access标志位，初始化位1，表示访问
mau['is_access'] = 1
# 将fp_mau_1和mau2月份的数据进行左合并
fp_mau_1 = fp_mau_1.merge(mau[mau.region_month=='2013-02'][['user_id', 'is_access']], how='left', on='user_id')
# # 将缺失值替换位0
fp_mau_1.loc[fp_mau_1.is_access.isnull(), 'is_access'] = 0
fp_mau_1.head()

region_month   user_id device  is_access
0      2013-01  10061580     FP        1.0
1      2013-01  10154440     FP        0.0
2      2013-01  10165615     FP        1.0
3      2013-01  10321356     FP        1.0
4      2013-01  10447112     FP        1.0

上表中的is_access属性，1表示1月份非智能手机用户在2月仍然有访问，0表示没有访问。

**1月份非智能手机用户在2月份继续用非智能手机访问情况**

In [559]:
# 给fp_mau_2添加一个is_fp标志位，初始化位1，表示非智能手机
fp_mau_2['is_fp'] = 1
# 将fp_mau_1和fp_mau_2进行左合并
fp_mau_1 = fp_mau_1.merge(fp_mau_2[['user_id', 'is_fp']], how='left', on='user_id')
# 将缺失值替换为0
fp_mau_1.loc[fp_mau_1.is_fp.isnull(), 'is_fp']= 0
fp_mau_1.head()

region_month   user_id device  is_access  is_fp
0      2013-01  10061580     FP        1.0    1.0
1      2013-01  10154440     FP        0.0    0.0
2      2013-01  10165615     FP        1.0    1.0
3      2013-01  10321356     FP        1.0    1.0
4      2013-01  10447112     FP        1.0    1.0

**1月份非智能手机用户在2月份换用智能手机访问情况**

In [560]:
#  给sp_mau_2添加一个is_sp标志位，初始化位1，表示智能手机
sp_mau_2['is_sp'] = 1
# fp_mau_1和sp_mau_2进行左合并
fp_mau_1 = fp_mau_1.merge(sp_mau_2[['user_id', 'is_sp']], how='left', on='user_id')
# 将缺失值替换为0
fp_mau_1.loc[fp_mau_1.is_sp.isnull(), 'is_sp'] = 0
fp_mau_1.head()

region_month   user_id device  is_access  is_fp  is_sp
0      2013-01  10061580     FP        1.0    1.0    0.0
1      2013-01  10154440     FP        0.0    0.0    0.0
2      2013-01  10165615     FP        1.0    1.0    0.0
3      2013-01  10321356     FP        1.0    1.0    0.0
4      2013-01  10447112     FP        1.0    1.0    0.0

此时，我们需要建立一个判别模型，用来判断哪些非智能手机用户下个月是会通过智能手机访问还是会流失。在这个过程中，下个月继续使用非智能手机访问的用户数据对于建立模型没有什么作用，将其过滤掉：

In [561]:
# 1月份使用非智能手机访问但是2月份停止访问的用户或2月份使用智能手机访问的用户
fp_mau_1 = fp_mau_1.loc[(fp_mau_1.is_access==0) | (fp_mau_1.is_sp==1)].copy()
fp_mau_1.head()

region_month   user_id device  is_access  is_fp  is_sp
1       2013-01  10154440     FP        0.0    0.0    0.0
7       2013-01  10528830     FP        0.0    0.0    0.0
20      2013-01   1163733     FP        1.0    0.0    1.0
21      2013-01  11727630     FP        0.0    0.0    0.0
43      2013-01  13401362     FP        1.0    0.0    1.0

这时的fp_mau_1就是可以用于逻辑回归分析的数据形式了。

#### 5.3.2  关于每天是否登录游戏的数据的整理

现在需要对每个用户每天的访问情况进行加工处理。原始dau的region_day属性记录了每个用户每天登录情况，但是其格式不符合我们的需求。

dau中的数据格式如下：

|region_day|user_id|
|---|---|
|2013-01-01|10061580|
|2013-01-02|10061580|
|······|······|
|2013-01-31|10061580|

我们需要的格式如下：

|user_id|1|2|······|31|
|---|---|---|---|---|
|10061580|1|1|······|0|
|······|······|······|······|······|

**对1月份非智能手机用户日访问数据进行加工**

In [562]:
# 提取1月份非智能手机用户数据
fp_dau_1 = dau[(dau.device=='FP') & (dau.region_month=='2013-01')][['user_id', 'region_day']]
fp_dau_1.head()

user_id  region_day
0  10061580  2013-01-01
1  10154440  2013-01-01
3  10165615  2013-01-01
4  10321356  2013-01-01
6  10447112  2013-01-01

In [563]:
# 将region_day属性转换为日期格式并提取day
fp_dau_1['region_day'] = pd.to_datetime(fp_dau_1.region_day)
# 提取region_day属性的day
fp_dau_1['region_day'] = fp_dau_1.region_day.dt.day
fp_dau_1.head()

user_id  region_day
0  10061580           1
1  10154440           1
3  10165615           1
4  10321356           1
6  10447112           1

In [564]:
#  按照user_id和region_day对数据进行聚合，并统计uer_id个数（作为是否登录标志）
fp_dau_1 = fp_dau_1.groupby(['user_id', 'region_day'])['user_id'].agg('count').unstack()
# 将缺失值替换为0
fp_dau_1.fillna(value=0, inplace=True)
fp_dau_1.head()

region_day   1    2    3    4    5    6    7    8    9    10 ...    22   23  \
user_id                                                      ...              
397286      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0 ...   1.0  1.0   
471341      1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
503874      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
512250      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0 ...   1.0  1.0   
513811      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   1.0  0.0   

region_day   24   25   26   27   28   29   30   31  
user_id                                             
397286      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
471341      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
503874      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
512250      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
513811      0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0  

[5 rows x 31 columns]

fp_dau_1中包含有1月份使用非智能手机在2月份继续使用非智能手机的日访问数据。所以在和fp_mau_1进行合并时，需要使用**内合并**。

In [565]:
# 将fp_dau_1和fp_mau_1数据进行内合并
fp_dau_1_cast = fp_dau_1.merge(fp_mau_1[['user_id', 'is_sp']], how='inner', on='user_id')
fp_dau_1_cast.head()

user_id    1    2    3    4    5    6    7    8    9  ...     23   24   25  \
0   471341  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  ...    0.0  0.0  0.0   
1   503874  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0  0.0  0.0   
2  1073544  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    1.0  1.0  1.0   
3  1073864  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0  0.0  0.0   
4  1163733  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    1.0  1.0  1.0   

    26   27   28   29   30   31  is_sp  
0  0.0  0.0  0.0  0.0  0.0  0.0    1.0  
1  0.0  0.0  0.0  0.0  0.0  0.0    0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0    0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0    0.0  
4  1.0  1.0  1.0  0.0  0.0  0.0    1.0  

[5 rows x 33 columns]

In [566]:
fp_dua_1_cast.is_sp.value_counts()

0.0    190
1.0     62
Name: is_sp, dtype: int64

经确认，标记为“0”的用户有190名，标记为“1”的用户有62名，共计252名。

## 6 验证是否能够建立模型

在本例中，基于1月份31天的访问数据来创建一个计算2月份账号迁转概率的模型。建立这样的预测模型的情况下，分为大体上两部分，建立模型和验证模型：
- **使用现有数据进行逻辑回归分析并建立预测模型。**
- **为了验证所得到预测模型的可信度，需要将模型的预测值和实际数据进行对比**

本例中，模型索要预测的时“下月账号迁转的概率”，预测值大于0.5时，表示下月用户账号发生迁移，而小于0.5时则表示用户流失。

### 6.1 使用逻辑回归来建立模型 

在本例中，自变量指的是过去1月（31天）的访问情况，但是这样变量的个数有些过多，为了得到更加合适的模型，需要按照某种标准对变量进行选择。选择变量的方法有很多种，一般经常使用一种名为AIC（赤池信息量准则）的指标。基于这个指标来选择变量并生成模型，如下所示：

这里选择X1day（1/1）、X4day（1/4）、X5day（1/5）、X7day（1/7）、X10day（1/10）、X13day（1/13）、X22day（1/22）、X29day(1/29)、X31day（1/31）9个变量。

**AIC具体用python如何操作，还没找到方法，先按照书籍手动选择如上9个特征**。

In [567]:
# 按照上述9个变量重构fp_dau_1_cast
re_fp_dau = fp_dau_1_cast.reindex(columns=['is_sp', 1, 4, 5, 7, 10, 13, 22, 29, 31])
re_fp_dau.rename(columns={1: 'X1day', 4: 'X4day', 5: 'X5day', 7: 'X7day', 10: 'X10day', 13: 'X13day', 22: 'X22day', 29: 'X29day', 31: 'X31day'}, inplace=True)
re_fp_dau.head()

is_sp  X1day  X4day  X5day  X7day  X10day  X13day  X22day  X29day  X31day
0    1.0    1.0    1.0    0.0    0.0     0.0     0.0     0.0     0.0     0.0
1    0.0    1.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0     0.0
2    0.0    0.0    0.0    0.0    0.0     1.0     0.0     0.0     0.0     0.0
3    0.0    0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0     0.0
4    1.0    1.0    0.0    0.0    0.0     0.0     0.0     1.0     0.0     0.0

**使用重构后的fp_dau_1_cast，即re_fp_dau来构建罗建回归模型**

In [568]:
# 划分训练集特征值和标签
X_train = re_fp_dau.values[:, 1:]
y_triain = re_fp_dau.values[:, 0]
# 初始化逻辑回归模型
clf = LogisticRegression(solver='liblinear', C=1.0, penalty='l1', tol=1e-6)
# 使用（X_train, y_train）来训练模型
clf.fit(X_train, y_triain)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l1',
                   random_state=None, solver='liblinear', tol=1e-06, verbose=0,
                   warm_start=False)

**关联模型系数和对应特征**

In [569]:
# 将回归系数和对应特征关联起来
coef_feature = pd.DataFrame({'coef': list(clf.coef_)[0], 'feature': re_fp_dau.iloc[:, 1:].columns})
coef_feature

coef feature
0  0.970285   X1day
1  1.090181   X4day
2 -0.004883   X5day
3  0.762553   X7day
4 -0.952864  X10day
5  0.488921  X13day
6  1.298988  X22day
7  1.557569  X29day
8  1.049979  X31day

通过观察，发现X5day和X10day的系数为负数（是一个负相关）。如果直观的解释，那就是”如果5日和10日没来访问的话，那么下月（2月）账号就会发生迁移“，意思有些莫名其妙。出现这种情况是由于自变量之间存在着相关性，把这种现象称为多重共线性。

当出现多重共线性时，上述基于系数的讨论就没有什么意义了。而且这种情况下得到的模型和不稳定，无法进行长期预测。一般来说此时就需要做进一步的变量选择以及增加交互作用等，以便得到再现性更强的模型。

然而，本例中，是否能够建立模型才是我们所关注的，所以暂时不需要进行上述工作。这次建立的模型并不一定要长期有效，只要能够在本例中使用即可，也就是所谓的一次性模型。

由于从一开始就没有包含正解的数据，因此就不用再拘泥于模型的精度或者是再现性。综上，建立的模型至少再本案例中不存在使用上的问题，可以用这个模型继续下面的工作。

### 6.2 使用模型进行预测

**用得到的模型来预测（训练集自身？）结果：**

In [570]:
# 在训练集上直接进行预测
prob = clf.predict_proba(X_train)
prob.shape

(252, 2)

predict_proba()函数的返回结果是每条记录根据模型拟合后各个标签（**标签经过升序排序**）发生的概率。

例如，本例中标签取值0和1，则返回结果则是每条记录经过模型拟合后其取值为0的概率和取值为1的概率。结果是形状为(252, 2)的数组。第一列为取值为0的概率，第二列为取值为1的概率。

In [571]:
# 取出取值为1的概率列
prob_1 = prob[:, 1:]
# 将该属性新增到fp_dau_1_cast中
fp_dau_1_cast['prob_1'] = prob_1
fp_dau_1_cast.head()

user_id    1    2    3    4    5    6    7    8    9    ...      24   25  \
0   471341  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0    ...     0.0  0.0   
1   503874  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0    ...     0.0  0.0   
2  1073544  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0    ...     1.0  1.0   
3  1073864  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0    ...     0.0  0.0   
4  1163733  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0    ...     1.0  1.0   

    26   27   28   29   30   31  is_sp    prob_1  
0  0.0  0.0  0.0  0.0  0.0  0.0    1.0  0.279683  
1  0.0  0.0  0.0  0.0  0.0  0.0    0.0  0.115453  
2  0.0  0.0  0.0  0.0  0.0  0.0    0.0  0.018718  
3  0.0  0.0  0.0  0.0  0.0  0.0    0.0  0.047133  
4  1.0  1.0  1.0  0.0  0.0  0.0    1.0  0.323611  

[5 rows x 34 columns]

根据prob_1属性来生成预测值（实际上可以用clf.predict()直接拿到预测值，这里为了学习clf.predict_proba()函数），当这个值大于0.5时取1（账号迁转），小于0.5时取0（用户流失）：

In [572]:
# fp_dau_1_cast添加pred属性
fp_dau_1_cast['pred'] = np.where(fp_dau_1_cast.prob_1>0.5, 1, 0)
fp_dau_1_cast.head()

user_id    1    2    3    4    5    6    7    8    9  ...    25   26   27  \
0   471341  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  0.0  0.0   
1   503874  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  0.0  0.0   
2  1073544  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   1.0  0.0  0.0   
3  1073864  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  0.0  0.0   
4  1163733  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   1.0  1.0  1.0   

    28   29   30   31  is_sp    prob_1  pred  
0  0.0  0.0  0.0  0.0    1.0  0.279683     0  
1  0.0  0.0  0.0  0.0    0.0  0.115453     0  
2  0.0  0.0  0.0  0.0    0.0  0.018718     0  
3  0.0  0.0  0.0  0.0    0.0  0.047133     0  
4  1.0  0.0  0.0  0.0    1.0  0.323611     0  

[5 rows x 35 columns]

到目前为止，收集了分析所需的每个用户账号的各种信息。
- 1月份用户的日访问情况：当前有访问设置为1，没有访问设置为0
- 计算is_sp属性的值：过滤掉1月份使用非智能手机在2月份继续使用非智能手机的用户
- 计算prob_1属性的值：用训练得到的模型在训练集上进行预测，并抽取标签取值为1（账号迁转）的概率
- 以0.5为基准，将prob_1属性离散为0（用户流失）和1（账号迁转）

这里使用最后一列属性pred来对模型的可信度进行验证。确认预测值和实际值之间差异的方法有很多种，本例使用的是交叉统计表。

In [573]:
# 生成实际值和预测值的交叉统计表
is_sp_cross_pred = pd.crosstab(index=fp_dau_1_cast.is_sp, 
                               columns=fp_dau_1_cast.pred, 
                               rownames=['实际值'], 
                               colnames=['预测值'], 
                               aggfunc='count',
                               values=fp_dau_1_cast.user_id, 
                               margins=True,
                               margins_name='总计'
                              )
is_sp_cross_pred.rename(index={0.0: '用户流失', 1.0: '账号迁转'}, columns={0: '用户流失', 1: '账号迁转'}, inplace=True)
is_sp_cross_pred

预测值   用户流失  账号迁转   总计
实际值                  
用户流失   185     5  190
账号迁转    24    38   62
总计     209    43  252

由上表可知：
- 模型预测为”用户流失“的用户有185名，而这185名用户确实不再来了。
- 模型预测为”用户流失“但实际上用户进行了”账号迁转“的用户有24名。
- 模型预测为”账号迁转“的用户有38名，而着38名用户确实进行了账号迁转。
- 模型预测为”账号迁转“但实际上”用户流失“的用户有5名。

预测的准确率为$\frac{185+38}{185+5+24+38}=88\%$

这个准确率比较高，基本可以认为得到了一个值得信赖的预测模型。然而在实际使用中，直接使用上述模型是不行的，还需要通过交叉验证的方法来做进一步验证。

### 6.3 根据预测结果推测用户群

由上述交叉表可知，该模型中预测值为”1“（账号迁转）而实际流失的用户有5名，可以认为这些人是因为账号迁转失败而流失的。

也就是说，根据过去访问情况推测出用户应该会在装好迁转设定后继续使用，但实际上这群流失的用户由于账号迁转失败而不再访问游戏了。

**查看预测为"账号迁转"实际也为"账号迁转"的38名用户的实际情况**

In [581]:
# 查看is_sp为1且pred为1用户情况
fp_dau_1_cast1 = fp_dau_1_cast.loc[(fp_dau_1_cast.is_sp==1.0) & (fp_dau_1_cast.pred==1)].copy()
# 按prob_1降序排序
fp_dau_1_cast1.sort_values(by='prob_1', ascending=False, inplace=True)
fp_dau_1_cast1.head()

user_id    1    2    3    4    5    6    7    8    9  ...    25   26  \
137  24791702  1.0  1.0  0.0  1.0  0.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   
136  24791702  1.0  1.0  0.0  1.0  0.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   
138  24900784  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   
134  24581383  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   
133  24581383  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   

      27   28   29   30   31  is_sp    prob_1  pred  
137  1.0  1.0  1.0  1.0  1.0    1.0  0.962999     1  
136  1.0  1.0  1.0  1.0  1.0    1.0  0.962999     1  
138  1.0  1.0  1.0  1.0  1.0    1.0  0.962825     1  
134  1.0  1.0  1.0  1.0  1.0    1.0  0.962825     1  
133  1.0  1.0  1.0  1.0  1.0    1.0  0.962825     1  

[5 rows x 35 columns]

上述结果出现的两条重复记录是因为该用户在2月份先使用非智能手机访问后，再使用智能手机访问。

数据中预测值和实际值都为”1“的用户确实仍在频繁的访问游戏。

**查看预测为"账号迁转"实际为"用户流失"的5名用户事实情况**

In [582]:
# 查看is_sp为0且pred为1用户情况
fp_dau_1_cast2 = fp_dau_1_cast.loc[(fp_dau_1_cast.is_sp==0) & (fp_dau_1_cast.pred==1)].copy()
# 按照prob_1降序排列
fp_dau_1_cast2.sort_values(by='prob_1', ascending=False, inplace=True)
fp_dau_1_cast2.head()

user_id    1    2    3    4    5    6    7    8    9  ...    25   26  \
203  43451947  1.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0  ...   1.0  0.0   
197  42276142  1.0  1.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0  ...   1.0  1.0   
194  41590801  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  0.0   
108  19432099  1.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0  1.0  ...   0.0  0.0   
27    6147878  1.0  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   

      27   28   29   30   31  is_sp    prob_1  pred  
203  0.0  1.0  1.0  0.0  0.0    0.0  0.847526     1  
197  1.0  1.0  1.0  0.0  0.0    0.0  0.808717     1  
194  0.0  0.0  1.0  0.0  1.0    0.0  0.710957     1  
108  0.0  0.0  0.0  0.0  0.0    0.0  0.657363     1  
27   0.0  0.0  0.0  0.0  0.0    0.0  0.656262     1  

[5 rows x 35 columns]

数据中”1“出现的次数很多，这表明大多数用户在1月份还是很频繁的访问游戏。这5名用户在这段时间内对游戏仍有很强烈的兴趣，不太可能因为兴趣变淡了而不再访问游戏。所以，这类用户很可能就是因为**账号迁转太麻烦而放弃了游戏**。

**查看预测为"用户流失"实际也为"用户流失"的185名用户实际情况**

In [585]:
# 查看is_sp为0且pred为0的用户情况
fp_dau_1_cast3 = fp_dau_1_cast.loc[(fp_dau_1_cast.is_sp==0) & (fp_dau_1_cast.pred==0)].copy()
# 按照prob_1降序排序
fp_dau_1_cast3.sort_values(by='prob_1', ascending=False, inplace=True)
fp_dau_1_cast3.head()

user_id    1    2    3    4    5    6    7    8    9  ...    25   26  \
234  56324381  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  0.0   
29    6658192  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  ...   0.0  0.0   
156  28315570  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  ...   0.0  0.0   
209  46285446  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  1.0  ...   1.0  1.0   
144  25483214  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0  ...   0.0  0.0   

      27   28   29   30   31  is_sp    prob_1  pred  
234  0.0  0.0  1.0  0.0  0.0    0.0  0.462584     0  
29   0.0  0.0  0.0  0.0  0.0    0.0  0.453048     0  
156  0.0  0.0  0.0  0.0  0.0    0.0  0.453048     0  
209  1.0  0.0  1.0  0.0  0.0    0.0  0.449587     0  
144  1.0  0.0  1.0  0.0  0.0    0.0  0.364950     0  

[5 rows x 35 columns]

由上表可知，这类用户已经不怎么来访问游戏了，他们对游戏的兴趣在逐渐的降低，因此也就慢慢的不再来访问了。

**通过上述内容可以看出，本例中建立的模型可信度较高，也就是说当初建立的假设“账号迁转设定失败而导致的用户流失”的影响较小，从实际人数上来看也只有5名用户。**

## 7 解决对策

整个分析过程如下：
- **事实：**
    - 非智能手机用户量的减少比智能手机用户量的增加要大
    - 当用户从非智能手机换到智能手机后，重新注册游戏账号的用户数量很少（从问卷调查中得出的结论）
- **假设：**
    - 从非智能手机更换到智能手机时，游戏账号的迁转设定失败导致了用户流失
- **基于分析方法和结果而采取的解决对策：**
    - 使用用户访问情况数据来建立“账号迁转”和“用户流失”的判别模型
        - **能够建立有效模型：**因游戏账号迁转设定失败而导致的用户流失的影响较小$\Rightarrow $改进面向智能手机新用户的商业宣传
        - **不能建立有效的模型：**因游戏账号迁转设定失败而导致的用户流失的影响较大$\Rightarrow $简化游戏账号的迁转设定

假设为“在智能手机上游戏账号迁转设定失败导致流失的用户很多”，而根据分析结果，因账号迁转设定失败而流失的用户占整体流失用户的比例很小。因此，可以改进面向智能手机新用户的商业宣传。